[자동차 경주게임] 무 미션 제출합니다.

veggie-garden opened the pull request:
안녕하세요.  

우선 제출이 늦은 점 죄송합니다. 저번 미션에서 하지 못한 모듈화를 하고 싶어 코드를 수정하다 보니 제출이 늦어졌습니다. 제 불찰입니다. 

지금 면접 기간이라 바쁘시다고 들었습니다. 그럼에도 귀한 시간 내어 코드 리뷰 해주셔서 감사합니다.  

많이 부족한 코드이지만, 열심히 짜봤습니다. 리뷰 받고 더욱 발전된 모습 보여드릴 수 있도록 노력하겠습니다.   

감사합니다!  

------------------------------------------------------

https://veggie-garden.github.io/javascript-racingcar-precourse/

이번 과제는 아래와 같은 목표를 두고 진행했습니다:

1. 모듈화
    - 모듈에 관한 공부 및 파일 분리
2. 모든 요구사항 충족
    - 기능 요구사항 
       - 기능 목록 모두 구현
    - 프로그래밍 요구사항
       - Car 객체 생성 및 사용
    - 과제 진행 요구사항
       - commit log 간결하나 알아보기 쉽게 적기 

## 모듈화
저번 미션에서는 모듈에 대한 이해도가 낮아 모듈화를 하지 못했습니다. 이번에는 객체를 사용해 모듈화를 진행하려 했으며, [3기 크루분의(zereight님) 코드](https://github.com/zereight/javascript-racingcar-precourse/tree/zereight)를 참고했습니다. 객체로 코드를 짜니, 자동차의 이름이라든지, 자동차가 몇 번을 전진했는지 등 자동차에 관한 모든 정보를 객체에 저장할 수 있어 필요할 때마다 정보를 꺼내 쓰고 저장하기 좋았습니다. 또한 모듈화로 파일들을 분리해두니, 원하는 함수를 원하는 곳에 불러 사용할 수 있어서 좋았습니다. 처음으로 객체를 코드에 사용했는데 왜 객체지향 프로그래밍을 해야 하는지 100% 이해하지 못했지만 조금은 알게 된 의미 있는 시간이었습니다.

## 모든 요구사항 충족 
기능 요구사항을 모두 구현하려고 노력했습니다. 그러나 이번에는 단순 구현에서만 멈추는 것이 아니라 모듈화와 파일 분리, 그리고 객체를 사용하고자 노력했습니다. 또한 프로그래밍 요구사항 - Car 객체를 생성하여 사용, 또한 지키려고 노력했습니다. 저번 과제를 진행하면서 아쉬웠던 점 하나는 commit 메세지가 일관성이 없고 헤더가 너무 길다는 점이었습니다. 그래서 이번에는 최대한 간결하고 어떤 변경 사항이 있었는지 알아보기 쉽게 작성하려 했습니다. 또한 공백을 잘 활용하여 가독성을 높이고자 했습니다. 공백을 포함하니 함수가 길어져 함수를 더 작게 쪼게야 했습니다. 함수와 변수의 역할을 최대한 들어내는 이름을 지으려 했습니다. 

## 결론
많이 미흡하고 부족한 코드이나 목표한 바를 이뤄서 뿌듯하고 기분이 좋습니다. 비록 아주 작은 성취일지라도, 성장할 수 있어 정말 기쁘고, 감사합니다. 쓴소리 많이 해주세요. 깐깐하게 봐주신다면 더욱 감사하겠습니다. 달게 받겠습니다. 감사합니다!

veggie-garden committed:
docs: add docs\README.md

veggie-garden committed:
chore: add eslint and prettier

veggie-garden committed:
style: add id into input and button

veggie-garden committed:
docs: update readme

veggie-garden committed:
feat: get car names

veggie-garden committed:
feat: check valid car name

veggie-garden committed:
feat: add display controls

veggie-garden committed:
chore: add constant

veggie-garden committed:
style(index.js): add id

veggie-garden committed:
chore: update constants

veggie-garden committed:
refactor: get car names from user

veggie-garden committed:
feat: prevent submit by enter

veggie-garden committed:
feat: get racing count and check validation

veggie-garden committed:
feat: start game

veggie-garden committed:
feat: print result of each turn

veggie-garden committed:
feat: print winner

veggie-garden committed:
docs: update readme

veggie-garden committed:
chore: typo correction

veggie-garden committed:
chore: create for github pages

veggie-garden committed:
fix: getWinner correction

veggie-garden committed:
Merge branch 'veggie-garden' of github.com:veggie-garden/javascript-racingcar-precourse into veggie-garden

bigsaigon333 commented on .nojekyll:
```diff
@@ -0,0 +1 @@
+
```
요건 무슨 파일일까요?! 
필요하지 않은데 에디터로 인해 자동으로 생성되거나 하면 .gitignore 에 추가하는 것도 좋을 것 같아요 :+1:

veggie-garden replied:
제 github.io는 Jekyll을 사용하고 있어서 github pages 를 생성할 때마다 테마가 적용됩니다. 그래서 Jekyll을 사용하지 않는다는 nojekyll 파일을 생성하면 테마가 적용되지 않아 이 파일을 생성한 것입니다!

bigsaigon333 replied:
그래서 `.nojekyll` 이 필요한거군요 ~ 이해하였습니다 :+1:

bigsaigon333 commented on package.json:
```diff
@@ -4,7 +4,13 @@
+    "eslint-config-prettier": "^8.3.0",
+    "eslint-plugin-import": "^2.25.3",
+    "eslint-plugin-prettier": "^4.0.0",
+    "prettier": "2.5.1"
```
Goooooooooood :+1:

bigsaigon333 replied:
lint-staged 로 eslint와 prettier 적용을 commit 전에 자동화하는 것도 추천드립니다 :+1:

bigsaigon333 replied:
<https://www.npmjs.com/package/lint-staged>

veggie-garden replied:
아하 저런 것도 있군요! 좋은 정보 감사드립니다. 바로 사용해보겠습니다.

bigsaigon333 commented on .eslintrc.json:
```diff
@@ -0,0 +1,18 @@
+{
+    "env": {
+        "browser": true,
+        "es2021": true
+    },
+    "plugins": ["prettier"],
+    "extends": [
+        "airbnb-base",
+        "plugin:prettier/recommended"
+    ],
+    "parserOptions": {
+        "ecmaVersion": "latest",
+        "sourceType": "module"
+    },
+    "rules": {
+        "prettier/prettier": "error"
+    }
+}
```
indent의 space 개수로 보아, 이 파일에는 prettier 가 적용되지 않은 것 같네요 :dizzy: 

`npx prettier --write .`  또는 package.json 에 script를 추가하는 방법 등을 통해서 
json 파일도 prettier 를 적용하면 더 좋지 않았을까 생각합니다 :smile:

veggie-garden replied:
미처 package.json 파일도 prettier를 적용해야 한다는 생각을 못 했네요. 수정하겠습니다! 그런데 package.json에 script를 추가한다는 말은 무슨 뜻인가요?

bigsaigon333 replied:
package.json 에 scripts 라는 프로퍼티를 추가하여 아래와 같이 정의해두고 사용하면 좋겠다는 뜻이었습니다  :smile: 

```js
scripts: {
  prettier: "prettier --write ."
}
```

<https://docs.npmjs.com/cli/v8/using-npm/scripts>

veggie-garden replied:
아하... 이런 것도 가능하군요! 좋은 정보 감사합니다😁

bigsaigon333 commented on index.html:
```diff
@@ -13,16 +13,16 @@ <h1>🏎️ 자동차 경주 게임</h1>
-      <h4>📄 실행 결과</h4>
+      <h4 id="result">📄 실행 결과</h4>
     </div>
     <script type="module" src="src/index.js"></script>
```
script 를 body 태그 가장 하단에 위치시키는 이유는 뭘까요?
type이 module 인 script와 아닌 script의 차이는 무엇일까요?
한번쯤 고민해보셨으면 좋겠습니다 :smile:

veggie-garden replied:
script 선언 위치에 따라서 차이가 생기는 것으로 알고 있습니다. body는 위에서 아래로 읽기 때문에, 랜더링이 완료된 후에 script가 실행되어 화면이 빠르게 뜨는 것이라고 배웠습니다. 그리고 script를 여러 개의 작은 파일들로 분리한 것을 module이라고 부르는데, 모듈은 다른 모듈 안에 있는 함수를 호출하는 등의 기능 공유가 가능합니다. 모듈은 자신만의 스코프를 가지고 있으며 이것이 일반 스크립트와의 차이라고 생각합니다.

bigsaigon333 replied:
예를 들어  A 라고 코드를 작성을 하셨고,  A라고 작성한 이유가 B라고 한다면,
A는 코드로 눈에 보이기 때문에 다른 사람들은 A라고 작성한 이유 B에 대해서 묻게 될꺼에요
그랬을 때 `B로 알고 있기 때문에 A로 작성하였습니다`, `B로 배웠기 때문에 A로 하였습니다` 보다
`B이기 때문에 A로 하였습니다` 라고 말씀하시는게 좋다고 생각합니다 :smile: 

자신이 틀릴 수 있다는 가능성을 인지하여 겸손(?)한 표현을 하는 것은 B의 결과로 나온 A에 대해서 가져야 할 자세이지,
B라는 본인의 생각에 가져야하는 것이 아니라고 생각해요. 
본인만의 생각이 정립되지 않고 코드A를 작성하였다면, 제대로 이해하지 못한채 인터넷에서 찾아서 갖다 썼다 라고 시인하는 것과 같다고 생각해요 :sweat_smile: 

표현을 너무 겸손하게 하시는거 같아서 저의 평소 생각을 좀 강하게 말씀드려봤습니다 :sweat_drops: 

--- 

추가 코멘트를 드리자면, 아래의 네가지 경우에 실행 시점이 정말 모두다 렌더링이 완료된 후인지 한번 확인해보시면 좋을 것 같아요 :smile: 

1. 그냥 script
2. script 에 defer 속성이 있는 경우
3. script에 async 속성이 있는 경우
4. script type="module" 인 경우

아래의 자료도 한번 참고해보세요~ :+1: 
<https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script#attr-type>

<https://javascript.info/script-async-defer>

veggie-garden replied:
아... 그렇게 느껴지겠군요... 그렇게 느낄 거라고 생각을 못 했어요. 앞으로 확실하게 제 생각을 말해야겠네요. 조언 감사합니다! 첨부하신 링크로 더 공부하겠습니다.

bigsaigon333 commented on src/car.js:
```diff
@@ -0,0 +1,15 @@
+import { generateRandomNumber } from "./utills.js";
```
```suggestion
import { generateRandomNumber } from "./utils.js";
```

veggie-garden replied:
아 이름에 오타가 있었네요! 수정했습니다. [수정 commit](https://github.com/veggie-garden/javascript-racingcar-precourse/commit/5ea2d15672771c0f75ccd95c5e78e5c8dfdf9d71#diff-25a6634263c1b1f6fc4697a04e2b9904ea4b042a89af59dc93ec1f5d44848a26)

bigsaigon333 commented on src/validation.js:
```diff
@@ -0,0 +1,73 @@
+  return false;
+};
+
+export function isCarNameValid(carNamesInput) {
```
함수 선언문으로 함수를 선언하시기도 하고, 함수 표현식으로 화살표 함수를 선언, 할당하기도 하시는데
어떨 때 함수 선언문이다 이런 나름의 규칙이 있으실까요?

veggie-garden replied:
파일 내에서 세분화된, 작은 역할을 맡는 함수를 화살표 함수로 선언했고, 그 작은 함수들을 실행하는 함수들 또는 `export default function` 함수들은 함수 선언문으로 선언했습니다. 저 나름의 규칙을 가지고 선언했는데, 혹시 이것도 코드 통일성을 헤치나요?

bigsaigon333 replied:
화살표 함수와 함수 선언문의 차이점은 무엇일까요?

자체 this binding이 있는지, 호이스팅이 되는지 ... 여러 가지가 있을 것 같아요. 

무 님께서 말씀하시는 기준은 화살표 함수와 함수 선언문의 차이와는 크게 상관이 없다고 생각되어요.  

반드시 한 쪽을 써야하는 이유가 있는 곳 외에는 하나로 통일해 쓰는게 일반적인 것 같아요 :smile:

veggie-garden replied:
말씀하신 것을 들으니, 저의 기준은 확실한 이유가 없어서 차라리 코드 통일성을 위해서 통일해서 쓰는 게 더 낫다고 생각이 드네요. 화살표 함수와 함수 선언문의 차이를 잘 몰랐는데 이참에 그 둘의 차이점도 공부해야겠어요!

bigsaigon333 replied:
:+1:

bigsaigon333 commented on src/index.js:
```diff
@@ -0,0 +1,98 @@
+  preventSubmitByEnterKey();
+  hideRacingCountForm();
+
+  carNamesButton.addEventListener("click", getCarNames);
+  racingCountButton.addEventListener("click", getRacingCount);
```
버튼을 click하는 것 뿐만 아 니라, input 에서 enter키를 치면 submit이 되는걸 유저들이 일반적으로 기대하는 것 같아요 :smile: 

조금만 수정하면 할 수 있을 것 같은데 어떻게 하는게 좋을까요? :dog:

veggie-garden replied:
저 또한 enter 키로 submit을 하는 것이 익숙해서 그렇게 만들려 했습니다. 그러나 예시에서 submit은 다 버튼을 클릭하는 걸로 되길래 일부로 `preventSubmitByEnterKey` 라는 함수를 만들어서 enter로 submit이 안되게 코드를 짰습니다. 만약 그 함수가 없으면 enter키로도 submit이 되는데 혹시 그것을 말씀하시는 건가요? 아니면 keydown을 말씀하시는 걸까요...?

bigsaigon333 replied:
`submit` event 를 의도한게 맞습니다 :smile: 

예시가 어떤건지 잘 모르겠지만... 예시에서 명시적으로 enter 키를 눌렀을 때 submit이 되지 않아야 된다고 적혀있지 않았다면, 
form 요소가 있고 그 안에 input 요소와 button 요소가 있는 현 상황에서는, `submit` event가 가장 상황에 적절하다고 생각해서 말씀드려봤습니다 :smile:

veggie-garden replied:
예시는 gif 파일을 의미한 것이었습니다. 말씀하신대로, 애초에 오직 버튼을 클릭해야지만 submit이 가능하게 만들고 싶었다면 form 요소가 없었겠죠. 아니면 조건에 명시되어 있다던지 했을텐데 그걸 생각 못했네요. enter로 submit 가능하게 수정했습니다!

bigsaigon333 replied:
:+1:

bigsaigon333 commented on src/index.js:
```diff
@@ -0,0 +1,98 @@
+  resultHeading.style.display = "block";
+};
+
+const getCarNames = (e) => {
```
함수명은 getCarNames 인데 CarNames 를 반환하지 않아서 어색한 것 같아요

그리고 get~ 이면 단순히 CarNames만 반환해야할 것 같은데 안에서 많은 로직이 수행되고 있는 것 같구요 :smiley:

bigsaigon333 replied:
event handler는 보통 get~ 으로 짓지 않고, handle~, on~ 이렇게 짓는게 일반적이긴 한 것 같아요

event handler 라는 특성을 떼버리고, 함수 자체가 무엇을 하는지를 한문장으로 보여줄 수 있다면 그렇게 하는게 제일 좋은거 같아요 :+1:

veggie-garden replied:
아하 함수명이 조금 부적절했군요. 보통 event handler는 handle~ 또는 on~ 이런식으로 이름을 짓는군요! event handler의 역할도 수행하지만 다른 일들도 하고 있어서, 말씀하신 바처럼 event handler 식의 이름을 짓지 않고, `createCarObjectIfNamesValid` 라는 이름을 지어주었습니다. 밑에 `getRacingCount` 함수도 `getCarNames` 함수와 동일한 문제를 가지고 있어서 이 함수 이름도 `startGameIfRacingCountValid` 이런 식으로 변경했습니다. 무슨 일을 하는 함수인지 이름을 보고 알 수 있으신가요? 그리고 이렇게 이름이 좀 길어져도 괜찮나요?

bigsaigon333 replied:
함수가 하는 일을 명백하게 나타낼 수 있다면 함수 이름이 길어지더라도 괜찮습니다 :smile: 

추가 코멘트 그냥 가볍게 남기자면, 

1. 함수 이름에 if, with 와 같은 말을 붙이는게 좋지 않다고 주장하는 개발자들도 있습니다. 
예를 들어 `startGameIfRacingCountValid` 대신 `startGame`을 쓰자고 주장하실 수도 있을 것 같아요
명백한 정도와 함수 이름의 길이의 적절한 밸런스를 찾아내는 건 경험이 쌓이면 잘 할 수 있지 않을까요~ :smile: 
저도 정답을 찾아낼만큼 경험이 없어서 이 부분은 전 pass 하겠습니다~~

2. `create`, `get`, `find` prefix 는 그냥 값을 반환할거라는 뉘앙스를 주는데, `createCarObjectIfNamesValid` 는 실제로 값을 반환하지 않고 있어요. 사실 이런게 굉장히 어려운 부분이죠... 추상화정도가 높아지면 함수명을 그래서 뭘로 해야되냐? .. 라는 질문으로 이어지는데... 저도 잘 모르겠습니다 ㅎㅎ 아마 그래서 handle~, on~ 이런 식으로 무미건조하게 함수명을 짓게되는건가 싶기도 해요 :smile:

veggie-garden replied:
함수 이름에 if나 with을 사용이 부적절하다는 의견이 있다니 가능하면 그런 이름을 짓는 걸 지양해야겠어요. `create` 도 반환값이 있을 거라는 뉘앙스를 풍기는군요. 아... 이름 짓기 한 번 어렵네요... 오늘도 이미지 속의 통계에 공감하게 됩니다... 이름은 `validCarNameObjects`와 `validRacingCountStartGame`으로 수정했습니다! 

![image](https://user-images.githubusercontent.com/63505022/146677519-04527292-78fd-4b1f-9ba9-85ac6573744c.png)

bigsaigon333 replied:
[읽기 좋은 코드가 좋은 코드다](http://www.kyobobook.co.kr/product/detailViewKor.laf?mallGb=KOR&ejkGb=KOR&barcode=9788979149142) 와 같은 책을 참고하는 것도 추천드립니다~! 
완전 JS 내용만 있지는 않지만, 단순히 저의 말을 듣고 따라하시는 것보다는 이 책을 믿는게 더 좋을거에요 :smile:

bigsaigon333 commented on src/index.js:
```diff
@@ -0,0 +1,98 @@
+  carNamesButton.addEventListener("click", getCarNames);
+  racingCountButton.addEventListener("click", getRacingCount);
+
+  carNamesButton.currentGame = gameSetting;
+  racingCountButton.currentGame = gameSetting;
```
`HTMLElement` 에 비표준 프로퍼티를 추가하는게 일반적인 방법은 아닌 것 같아요 :smile: 

꼭 이렇게 해야한다고 생각했던 이유가 있을까요? 고민했던 내용을 공유하면서 계속 이야기 해나가봐요 :smile:

veggie-garden replied:
아하 이게 일반적인 방법은 아니군요. 왜 저런 식으로 사용했나면, `carNameButton`이 클릭되면 `createCarObjectIfNamesValid`가 실행됩니다. 그러나 `createCarObjectIfNamesValid`이 실행되기 위해서 `gameSetting`이 필요한데, `carNamesButton.addEventListener("click", createCarObjectIfNamesValid(gameSetting));` 이런 식으로 작성하면, 유저가 게임을 시작하기 이전에 게임이 실행되어, 아무런 이름이 입력되지 않은체 올바른 이름인지 체크하는 함수가 실행되어서 alert이 뜹니다. 그래서 이걸 해결하고자 다른 분들의 코드를 살피다가 zereight 님의 코드를 따라했습니다. 이게 일반적인 방법이 아니라면, 보통 어떤 식으로 `createCarObjectIfNamesValid`에 `gameSetting`을 전달하나요?

veggie-garden replied:
![Animation](https://user-images.githubusercontent.com/63505022/146514370-8a6acc0b-ca33-475d-b9a6-d94a71aebb19.gif)

veggie-garden replied:
수정했습니다! 비표준 프로퍼티를 사용하는 대신 객체를 바로 불러오는 식으로 수정하니 위의 문제도 해결되었습니다. 저런 부분까지 하나하나 다 짚어 주셔서 덕분에 객체에 대한 이해도가 한결 상승했습니다. 아직도 많이 부족하지만요...ㅎㅎ 감사합니다!

bigsaigon333 replied:
말씀하신대로, JS의 클로저로 인해서 그냥 외부 변수들을 내부 함수내에서 접근이 가능한 점을 이용하면 좋은 것 같아요~~ :+1:

bigsaigon333 commented on src/car.js:
```diff
@@ -0,0 +1,15 @@
+  this.forwardCount = 0;
+}
+
+Car.forwardCar = (car) => {
+  const randomNumber = generateRandomNumber();
+  if (randomNumber > 3) {
+    car.forwardCount += 1;
+  }
+
+  return car;
+};
```
static method 로 구현하셨네요 :smile: 

이렇게 하신 이유가 있으실까요? 

car 라는 객체를 매개변수로 전달받아야 하기 때문에, static 이 아닌 그냥 메서드로 구현해도 괜찮을 것 같긴 해서 여쭤봅니다~

veggie-garden replied:
메서드로 구현할 수 있겠네요! 저렇게 짠 것은 Car 객체하고 RacingCarGame은 다른 객체니까 저런 식으로 분리해야 하지 않나 싶기도 하고 참고한 코드가 저렇게 되어 있어서 저렇게 짰습니다. 그렇게 대단한 이유는 없어요...😅 말씀하신대로 메서드로도 구현하는 방식으로 수정했습니다! 혹시 static method보다 일반 메서드로 구현하는 것을 추천하시는 이유가 있을까요?

bigsaigon333 replied:
일반 메서드로 구현한다는 건 아래와 같은 의미였습니다 :smile: 

```js
function Car(name) {
  this.carName = name;
  this.forwardCount = 0;
}

Car.prototype.moveForward = function() {
   const randomNumber = generateRandomNumber();
    if (randomNumber > 3) {
      this.forwardCount += 1;
    }
}

```

class를 사용한다면 아래와 같이 작성할 수 있겠네요

```js
class Car {
  constructor() {
    this.carName = name;
    this.forwardCount = 0;
  }

  moveForward() {
   const randomNumber = generateRandomNumber();
    if (randomNumber > 3) {
      this.forwardCount += 1;
    }
}
```

그러면 아래와 같이 racingCarGame을 작성할 수 있겠네요

```js
const racingCarGame = {
  cars: [],
  turnOfGame: 0,
  race() {
    for(const car of this.cars) {
      car.moveForward();
    }
  },
```

아래에 대해서도 한번 고민해보시면 좋습니다 :smile:
1. 메서드와 일반 함수의 차이는 무엇일까요?
2. 생성자 함수 + prototype 과 class 의 차이는 무엇일까요?

프로토타입을 자세하게 알아야 JS를 제대로 아는거라고 할 수 있습니다. 
쉬운 주제는 아니니 시간 날 때마다 학습하시면 좋을 것 같습니다~

veggie-garden replied:
공부할 주제를 주셔서 감사합니다! 한 번 공부해보겠습니다. `for(const car of this.cars) {}` 를 사용하니 eslint에서 사용하지 말라 하여 forEach 문으로 바꿨습니다.

bigsaigon333 replied:
저도 eslint 를 굉장히 좋아하는데요 (나를 가르친건 8할이 eslint이다..) 
eslint에서 왜 for-of 를 사용하지 말라고 하는건지도 한번 고민해보시면 좋습니다 :smile: 
iterable-iterator protocol, downlevelIteration, transpile, polyfill 등을 참고 키워드로 남겨드릴게요 :dog: 

그리고 eslint에서 왜 사용하지 말라는건지에 대해 완벽하게 이해가 되고 나면, 
이를 무작정 따르는게 아니라 자신만의 기준을 세우는 것이 좋습니다.
eslint 는 개발자의 생산성을 높이는 도구이므로, eslint 때문에 생산성을 떨어뜨리는 일은 없어야겠죠 :smile:

bigsaigon333 commented on src/index.js:
```diff
@@ -0,0 +1,98 @@
+import { getWinner, printResultOneTurn, printWinner } from "./utills.js";
+import { isCarNameValid, isRacingCountValid } from "./validation.js";
+
+export default function RacingCarGame() {
+  this.cars = [];
+  this.turnOfGame = 0;
+  this.race = () => {
+    for (let i = 0; i < this.cars.length; i += 1) {
+      this.cars[i] = Car.forwardCar(this.cars[i]);
+    }
+  };
+}
+
+RacingCarGame.startGame = (gameSetting) => {
+  while (gameSetting.turnOfGame > 0) {
+    gameSetting.race();
+    gameSetting.turnOfGame -= 1;
+    printResultOneTurn(gameSetting.cars);
+  }
+  printWinner(getWinner(gameSetting.cars));
+};
```
그냥 단일 객체로 만들면 어떨까요?

기계적으로 바꾼다면 아래와 같이 되겠네요 :smile: 

함수를 선언하고, 거기에 프로퍼티를 추가하는 방법을 사용할 필요가 굳이 있을까 라는 생각이 들어서 제안해봅니다 :dog: 

```js
const racingCarGame = {
  cars: [],
  turnOfGame: 0,
  race() {
    for (let i = 0; i < this.cars.length; i += 1) {
      this.cars[i] = Car.forwardCar(this.cars[i]);
    }
  },
  startGame() {
    while (this.turnOfGame > 0) {
      this.race();
      this.turnOfGame -= 1;
      printResultOneTurn(this.cars);
    }
    printWinner(getWinner(this.cars));
  },
};

```

veggie-garden replied:
아하 이런 식으로 하는게 코드가 더 깔끔하네요! 수정하겠습니다.

veggie-garden replied:
이런 식으로 하면 줄이 15줄이 넘어가는데 괜찮은 건가요?

bigsaigon333 replied:
15줄은 생각을 못했네요 ㅎㅎ
15줄을 맞춘다면, race랑 startGame을 함수로 추출해내거나 해야할거 같아요 :dog: 

근데 15줄 제약은 각 함수의 길이에 대한 거 아닌가요? 객체 리터럴의 길이에 대한 제약은 아닐거 같은데요 :dizzy:

veggie-garden replied:
> - 함수(또는 메소드)의 길이가 15라인을 넘어가지 않도록 구현한다.
>    - 함수(또는 메소드)가 한 가지 일만 잘 하도록 구현한다.  

함수만 그러면 되는 것이었네요...ㅎㅎ

bigsaigon333 commented on src/index.js:
```diff
@@ -0,0 +1,98 @@
+      e.preventDefault();
+    }
+  });
+};
```
90~91 행 리뷰를 참고해주세요 :smile:

bigsaigon333 commented on src/utills.js:
```diff
@@ -0,0 +1,47 @@
+  winnerSpan.innerText = winnerList.join(", ");
+
+  app.appendChild(resultSpan);
+  app.appendChild(winnerSpan);
```
```suggestion
  app.append(resultSpan, winnerSpan);
```

이렇게 쓸수도 있습니다 :smile:

veggie-garden replied:
앗 그렇군요! 수정했습니다.

bigsaigon333 replied:
추가 코멘트를 드리자면 `appendChild`와 `append`의 차이에 대해서도 학습해보시면 좋을 거 같아요. 단순히 인자를 몇개 받느냐도 있지만, 어떤 걸 인자로 받을 수 있는지도 다르답니다 :smile:

bigsaigon333 commented on src/utills.js:
```diff
@@ -0,0 +1,47 @@
+  app.appendChild(resultAreaP);
+};
+
+export const getWinner = (cars) => {
+  let winnerList = [];
+  let largestForwardCount = 0;
+
+  for (let i = 0; i < cars.length; i += 1) {
+    if (cars[i].forwardCount > largestForwardCount) {
+      largestForwardCount = cars[i].forwardCount;
+      winnerList = [cars[i].carName];
+    } else if (cars[i].forwardCount === largestForwardCount) {
+      winnerList.push(cars[i].carName);
+    }
+  }
+
+  return winnerList;
+};
```
`largestForwardCount` 를 찾는 로직과 
`car.forwardCount === largestForwardCount` 를 찾는 로직이 섞여 있네요

프론트엔드에서는 일반적으로 엄청나게 큰 숫자를 대상으로 한 계산을 하지 않기 때문에, 
성능 최적화보다는 가독성에 더 초점을 맞춰서 코드를 작성하시는게 좋습니다 :smile: 


```suggestion
export const getWinner = (cars) => {
  const largestForwardCount = Math.max(...cars.map(({forwardCount}) => forwardCount));
  // const largestForwardCount = cars.reduce((max, {forwardCount}) => Math.max(max, forwardCount), 0);
  const winnerList = cars.filter(({forwardCount}) => forwardCount === largestForwardCount);
  
  return winnerList;
};
```

veggie-garden replied:
아하 이렇게 활용할 수도 있군요..! 수정하겠습니다.

bigsaigon333 commented on src/validation.js:
```diff
@@ -0,0 +1,73 @@
+      alert(VALID_ALERTS.NAME_DUPLICATED);
+      return true;
+    }
+  }
+
+  return false;
+};
+
+const isCarNameEnough = (carNamesArr) => {
+  if (carNamesArr.length <= NUMS.MIN_LENGTH) {
+    alert(VALID_ALERTS.NAME_NOT_ENOUGH);
+    return true;
+  }
+
+  return false;
+};
+
+export function isCarNameValid(carNamesInput) {
+  const carNamesArr = divideCarNames(carNamesInput);
+
+  if (
+    isCarNameInputSpaces(carNamesInput) ||
+    isCarNameInputLength(carNamesArr) ||
+    isCarNameDuplicated(carNamesArr) ||
+    isCarNameEnough(carNamesArr)
+  ) {
+    return false;
+  }
+
+  return carNamesArr;
+}
+
+export function isRacingCountValid(count) {
+  if (count < NUMS.COUNT_MIN) {
+    alert(VALID_ALERTS.COUNT_NOT_VALID);
+    return false;
+  }
+
+  return true;
+}
```
밑의 글은 꽤 주관적인 저의 생각인데, 무님의 생각도 들려주세요 :smile: 

1. 각 `is~~valid` 함수가 alert 까지 호출하고 있는데,  `is~~valid` 함수는 단순시 boolean 값을 반환하는게 좋다고 생각해요

2. 저라면 validation 로직을 이렇게 작성할 것 같아요.
   1. `isCarNameValid` 함수명을 `getCarNameValidationMessage` 로 변경
   2.  각 is~~valid 함수는 boolean만 단순하게 반환
   3. `getCarNameValidationMessage` 함수에서는 `is~~valid` 함수의 반환값에 따라 validationMessage를 설정하고 이를 즉시 return. valid하면 `""` 또는 `null` 을 return
   4. `getCarNameValidationMessage` 함수를 호출하는 쪽에서 validationMessage에 따라 `alert`를 호출

3. 함수가 하나의 역할만 수행하게끔 작성하고, 자신의 역할이 아닌 것들은 호출하는 쪽에서 수행하는 것이 좋다고 생각합니다 :smile:

veggie-garden replied:
저도 boolean 함수는 boolean 값만 반환하는 것이 좋다는 조언을 본 뒤, 저도 그것이 맞다고 생각합니다. 그러려면 boolean 값을 판단하는 함수 하나, alert을 호출하는 함수 하나로 분리를 해야겠네요. 함수가 하나의 역할을 하게 만드려고 노력하는데 쉽지 않네요. 조언 감사합니다. 수정해보겠습니다.

bigsaigon333 commented on src/index.js:
```diff
@@ -0,0 +1,98 @@
+  resultHeading.style.display = "block";
+};
+
+const getCarNames = (e) => {
+  const game = e.currentTarget.currentGame;
+  const button = e.currentTarget;
+  const carNamesInput = document.getElementById("car-names-input");
+  const validCarNames = isCarNameValid(carNamesInput.value);
+  e.preventDefault();
+
+  if (validCarNames) {
+    game.cars = validCarNames.map((carName) => new Car(carName));
+    button.disabled = true;
+    showRacingCountForm();
+  }
+};
+
+const getRacingCount = (e) => {
+  const game = e.currentTarget.currentGame;
+  const button = e.currentTarget;
+  const racingCountInput = document.getElementById("racing-count-input");
+  e.preventDefault();
+
+  if (isRacingCountValid(racingCountInput.value)) {
+    game.turnOfGame = Number(racingCountInput.value);
+    button.disabled = true;
+    showResultHeading();
+    RacingCarGame.startGame(game);
+  }
+};
```
함수가 여러 역할을 할 때, 특히 이벤트 핸들러에서 많은 로직이 수행될 때 이렇게 생각해보면 좋은 것 같아요.

1. 상태가 무엇인지?
2. 상태가 변경되었을 때 다른 상태에 어떻게 영향을 주는지?
3. 상태가 변경되면 이를 UI로 어떻게 나타낼 것인지?

위의 예시를 적용해보면, 
- `#racing-count-submit` 이 클릭되면 `getRacingCount`가 호출이 되네요
- `#racing-count-input`이 가지고 있는 value가 상태인 것 같아요
-  상태가 최초에 0에서 유저가 입력한 값으로 변경이 되었어요
- 이는 `game.turnOfGame` 이라는 다른 상태에 영향을 주는 것 같아요. 그래서 업데이트를 했어요
- 그리고 게임도 시작해야 할 것 같아요 -> `RacingCarGame.startGame(game)`
- UI 적으로는 `#racing-count-submit`를 disabled 시키고, resultHeading 도 보여줘야해요

작성하면서 제가 든 생각은, 로직이 섞여있다는 생각도 들고, 
game이 시작된 이후의 UI 를 어떻게 처리하는지 모르겠다는 생각도 들었어요. 
아마 RacingCarGame.startGame 에서 UI 변경도 하고 있는거겠죠..? 
UI 변경과 상태 변경의 분리가 잘 안되어있다는 생각이 결론적으로 들어요 :dog: 

이런 식으로 한 번 분리해보는 것도 좋을것 같아요 :+1:

veggie-garden replied:
확실히 여러가지 로직을 수행하다보니 읽는 사람이 이해하기 어렵네요. 한 번 분리해보겠습니다!

bigsaigon333 commented on src/index.js:
```diff
@@ -0,0 +1,98 @@
+  const racingCountInput = document.getElementById("racing-count-input");
+  e.preventDefault();
+
+  if (isRacingCountValid(racingCountInput.value)) {
+    game.turnOfGame = Number(racingCountInput.value);
```
`HTMLInputElement` 에는 `valueAsNumber` 라는 프로퍼티도  있습니다 :smile: 

<https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement>

veggie-garden replied:
아 그런 프로퍼티도 있군요...! 좋은 정보 감사합니다. 앞으로 잘 활용하겠습니다!

bigsaigon333 reviewed (changes requested):
무님 안녕하세요~ 자동차 경주게임 구현하신다고 고생하셨습니다 :+1: 

배포를 해놓으셔서 실제로 테스트해보기 편해서 좋았습니다! 배려심 짱짱 :heart_eyes_cat: 

리뷰어 단톡방에서 제가 무님꺼 리뷰 하겠다고 굉장히 일찍 호기롭게 말했었는데, 리뷰를 늦게 남겨서 죄송해요 :sob: 

제가 공통적으로 리뷰이분들께 말씀드리는 사항 몇가지 말씀드릴게요.

1. 너무 처음부터 100% 다 이해하려고 안하셔도 돼요. 고민을 해보는게 중요한거고, 고민했던 내용을 개인 노션 같은 곳에 잘 정리해두시면 나중에 굉장히 도움 될거라고 생각합니다! 
(제가 그렇게 못했기 때문에 드리는 말씀...)
2. 리뷰는 저의 주관이 반영되기 때문에 반드시 반영하지 않으셔도 됩니다 :smile: 
그냥 개발자 대 개발자로서 얘기를 나눈다고 생각해주시면 좋을 거 같아요. 
코드를 작성한 사람이 코드에 대해서 제일 잘 알기 때문에, 고민했던 내용들을 저에게 말씀해주시면 더 좋은 얘기 이어나갈 수 있을 것 같아요.

리뷰를 다 반영(?) 하시고 나면 리뷰 요청 버튼을 다시 눌러주세요 :smile:

bigsaigon333 commented:
> 모듈화
저번 미션에서는 모듈에 대한 이해도가 낮아 모듈화를 하지 못했습니다. 이번에는 객체를 사용해 모듈화를 진행하려 했으며, 3기 크루분의(zereight님) 코드를 참고했습니다. 객체로 코드를 짜니, 자동차의 이름이라든지, 자동차가 몇 번을 전진했는지 등 자동차에 관한 모든 정보를 객체에 저장할 수 있어 필요할 때마다 정보를 꺼내 쓰고 저장하기 좋았습니다. 또한 모듈화로 파일들을 분리해두니, 원하는 함수를 원하는 곳에 불러 사용할 수 있어서 좋았습니다. 처음으로 객체를 코드에 사용했는데 왜 객체지향 프로그래밍을 해야 하는지 100% 이해하지 못했지만 조금은 알게 된 의미 있는 시간이었습니다.

음... `Car.forwardCar` 를 작성한게 zereight 의 코드를 참고하여서 나왔던 거군요..

본인의 코드에 대해서는 본인만의 이유가 반드시 있어야 합니다. 
제가 리뷰에도 남겼지만, `Car.forwardCar` 에 대한 무님의 생각을 남겨주세요 :smiley_cat: 
왜 그렇게 했는지, 다른 방법은 무엇이 있는지, 왜 이게 현 상황에서 최선이라고 생각한 것인지 한번 고민하는 계기가 되었으면 좋겠습니다.

그리고 객체를 사용한다고 다 객체지향 프로그래밍(OOP)은 아닙니다. 역할과 책임의 분리는 OOP에서만 하는 것도 아니구요 :smile:

bigsaigon333 commented:
> 모든 요구사항 충족
기능 요구사항을 모두 구현하려고 노력했습니다. 그러나 이번에는 단순 구현에서만 멈추는 것이 아니라 모듈화와 파일 분리, 그리고 객체를 사용하고자 노력했습니다. 또한 프로그래밍 요구사항 - Car 객체를 생성하여 사용, 또한 지키려고 노력했습니다. 저번 과제를 진행하면서 아쉬웠던 점 하나는 commit 메세지가 일관성이 없고 헤더가 너무 길다는 점이었습니다. 그래서 이번에는 최대한 간결하고 어떤 변경 사항이 있었는지 알아보기 쉽게 작성하려 했습니다. 또한 공백을 잘 활용하여 가독성을 높이고자 했습니다. 공백을 포함하니 함수가 길어져 함수를 더 작게 쪼게야 했습니다. 함수와 변수의 역할을 최대한 들어내는 이름을 지으려 했습니다.

commit 메세지 관련해서 제 생각을 말씀드리면, 리뷰어는 commit 메세지를 보지는 않는 것 같아요. 
우테코에서 리뷰받으면서, 리뷰를 직접하면서도 그랬어요. 
팀프로젝트를 진행하면 PR 단위로 develop 또는 main 브랜치에 merge 되기 때문에, 
PR내의 각 commit 메세지는 신경쓰지 않는게 일반적이었던 것 같아요. 

개발을 하면서 문제가 생겨서 전에 코드를 확인할 때 전에 commit 메세지를 확인하는게 일반적이기 때문에, 
본인이 개발하면서 편하고 일관된 방법으로 commit 메세지를 작성하는게 중요하다고 생각해요 :+1:

veggie-garden committed:
chore: change name to utils.js

veggie-garden committed:
refactor: rename getCarNames and getRacingCount function

veggie-garden committed:
chore: add lint-staged

veggie-garden committed:
refactor: refactor racingCarGame

veggie-garden committed:
chore: make two appendChild to one

veggie-garden committed:
refactor: getWinner and printWinner

veggie-garden committed:
refactor: return boolean value

veggie-garden committed:
refactor: separate getCarNames and getRacingCount into small functions

veggie-garden committed:
refactor: change forwardCar static method to method

veggie-garden committed:
chore: update eslintrc setting

veggie-garden commented:
안녕하세요, 동동님!  

우선 리뷰 받은 것 수정이 늦어서 죄송합니다. 상황을 설명하자면 우선, 제가 코드 리뷰 남겨주셨다는 걸 늦게 확인했고, 이번 주에 일정이 있어 그 일부터 해결하느라 리뷰 반영이 늦어졌습니다... 죄송합니다...   

최대한 리뷰 남겨주신 걸 적용해보려고 노력했는데 의도하신 바대로 잘했는지 모르겠네요. 

> 본인의 코드에 대해서는 본인만의 이유가 반드시 있어야 합니다.

마음에 새기도록 하겠습니다! 

> 왜 그렇게 했는지, 다른 방법은 무엇이 있는지, 왜 이게 현 상황에서 최선이라고 생각한 것인지 한번 고민하는 계기가 되었으면 좋겠습니다.

리뷰 받기 이전에는 완성하기에 급급해서 좋은 코드에 대한 고민이 부족했습니다. 그러나 굉장히 꼼꼼하게 코드 리뷰를 해주셔서 덕분에 저도 왜 제가 저렇게 짰는지, 말씀해주신 대로 더 좋은 방식은 없는지 고민하는 좋은 기회가 되었습니다.  

> 그리고 객체를 사용한다고 다 객체지향 프로그래밍(OOP)은 아닙니다. 역할과 책임의 분리는 OOP에서만 하는 것도 아니구요 😄 

아, 그렇군요! 코딩을 시작한지 얼마되지 않은 초짜라 아직 이런 부분에 대한 지식이 부족합니다. 알려주신 것들을 토대로 열심히 공부하도록 해야겠어요!

> 개발을 하면서 문제가 생겨서 전에 코드를 확인할 때 전에 commit 메세지를 확인하는게 일반적이기 때문에,
본인이 개발하면서 편하고 일관된 방법으로 commit 메세지를 작성하는게 중요하다고 생각해요 👍

조언 감사드립니다! commit message convention을 안 지 얼마 되지 않아 제 스타일을 찾는 중입니다. 조언 잘 적용해보도록 하겠습니다. 

다시 한 번 수정이 늦어서 죄송합니다. 바쁜 시간 쪼개가면서 정성스런 코드 리뷰 남겨주셔서 감사합니다!

bigsaigon333 commented on package.json:
```diff
@@ -24,5 +33,8 @@
     "node": ">=14.0.0"
+  },
+  "lint-staged": {
+    "*.js": "eslint --cache --fix"
```
:+1:

bigsaigon333 reviewed (approved):
고생하셨습니다~ 

추가 코멘트 몇개 남겼습니다! 부담갖지 마시고 가볍게 봐주시면 좋을거 같아요 :smile: 

만약 계속 코멘트 남겨주시면 저도 계속 코멘트 남기도록 하겠습니다 :+1:

veggie-garden committed:
feat: user can submit value by enter key

veggie-garden committed:
refactor: change function name

change createCarObjectIfNamesValid to validCarNameObjects, startGameIfRacingCountValid to validRacingCountStartGame

veggie-garden committed:
refactor: use method racingCarGame.race()

veggie-garden committed:
chore: change spaces of *.json

veggie-garden commented:
동동님도 고생 많으셨습니다! 

덕분에 많이 배웠습니다. 시간 내주셔서 감사합니다. 주신 링크들과 질문들 잘 공부하고 고민하겠습니다. 면접 잘 보셨기를 바라며, 요즘 확진자가 급증하고 있는데 코로나 조심하시고, 또 날씨가 부쩍 추워졌는데 건강 유의하세요. 

연말 잘 보내시고 메리 크리스마스입니다. 감사합니다☺️

bigsaigon333 commented:
@veggie-garden  님도 미션 수행하시느라 고생 많으셨습니다~ :+1: 
코멘트 남겨주신 부분에 대한 답글도 달아두었으니 참고해주세요 

연말연시 좋은 시간 보내시길 바랄게요 :smile: 
다음에 뵈어요~~~
